### PR TITLE
Add MetaWithOptions and requireNode callback

### PR DESCRIPTION
**Pull request by pottedmeat**
**Type:** feature

The following has been addressed in the PR:

* [ ] There is a related issue
* [ ] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [ ] Unit or Functional tests are included in the PR

**Description:**

* Adds an optional callback to `requireNode`
* Adds Meta/option bundling

### CODE REVIEW

**Comment by agubler** [comment]
@pottedmeat Would like to see an issue for bug/feature enhancement changes if possible. It would help us understand the goal of the changes.

File: src/WidgetBase.ts [hunk 1]
```diff
@@ -269,6 +284,25 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> extends E
 		return node;
 	}
 
+	@afterRender()
+	protected notifyMetaKeys(node: DNode | DNode[]): DNode | DNode[] {
```
Comment (agubler) [inline src/WidgetBase.ts]: Would like to try and avoid using `afterRenders` in `WidgetBase` after analysing the results from `js-framework-benchmark`

File: src/meta/Base.ts [hunk 2]
```diff
@@ -1,38 +1,61 @@
...
-import Set from '@dojo/shim/Set';
-import { WidgetMetaProperties } from '../interfaces';
+import {
+	WidgetMetaOptions,
+	WidgetMetaProperties,
+	WidgetMetaRequiredNodeCallback
+} from './interfaces';
 
-export class Base {
+export abstract class Base<T, O extends WidgetMetaOptions = WidgetMetaOptions> {
```
Comment (agubler) [inline src/meta/Base.ts]: I don't understand the first generic?

File: src/meta/MetaWithOptions.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import MetaBase from './Base';
+import { WidgetMetaConstructor, WidgetMetaOptions } from './interfaces';
+
+function MetaWithOptions<T extends MetaBase<any>, O extends WidgetMetaOptions>(MetaType: WidgetMetaConstructor<T, O>, options: O): MetaWithOptionsRef<T, O> {
+	return new MetaWithOptionsRef(MetaType, options);
+}
+
+export class MetaWithOptionsRef<T extends MetaBase<any>, O extends WidgetMetaOptions> {
+	private MetaType: WidgetMetaConstructor<T, O>;
```
Comment (agubler) [inline src/meta/MetaWithOptions.ts]: underscores for privates

File: src/WidgetBase.ts [hunk 4]
```diff
@@ -206,15 +211,25 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> extends E
...
 
-	protected meta<T extends MetaBase>(MetaType: WidgetMetaConstructor<T>): T {
-		let cached = this._metaMap.get(MetaType);
+	protected meta<T extends MetaBase<any>>(MetaOrMetaWithOptions: WidgetMetaConstructor<T> | MetaWithOptionsRef<T, any>): T {
+		let cached = this._metaMap.get(MetaOrMetaWithOptions);
 		if (!cached) {
+			let options: WidgetMetaOptions;
+			let MetaType: WidgetMetaConstructor<T>;
+			if (MetaOrMetaWithOptions instanceof MetaWithOptionsRef) {
+				options = (<any> MetaOrMetaWithOptions).options;
```
Comment (agubler) [inline src/WidgetBase.ts]: this feels like a hack

**Comment by kitsonk** [comment]
Also, as a general comment, I don't know if it calls for a slightly different approach, because in trying to create a meta manager for gestures (#571) I am realising it might just be easier to be specific in the `get()` API to pass options instead of the meta registration.

For example, I am thinking it will be something like this...

```ts
class FooWidget extends WidgetBase {
  render() {
    const scrollResult = this.meta(Gestures).get('root', { type: 'scroll' });
    const dragResult = this.meta(Gestures).get('draghandle', { type: 'drag' });
  }
}
```

This means the meta `Base` would stay the same without having to worry about abstract handling of options which isn't meaningful for every manager and makes it easier to provide different types of results which are totally under the control of manager instance, versus having potentially multiple managers per widget.

File: src/WidgetBase.ts [hunk 5]
```diff
@@ -269,6 +284,25 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> extends E
...
+			decorate(node, (node: any) => {
+				const { properties = {} }: { properties: { key?: string } } = node;
+				if (properties.key) {
+					keys.push(properties.key);
+				}
+			}, (node: DNode) => {
+				return isHNode(node) || isWNode(node);
+			});
+			this._metaMap.forEach((meta) => {
+				meta.onKeysUpdated(keys);
```
Comment (agubler) [inline src/WidgetBase.ts]: Doesn't feel like `onKeysUpdated` should be on the public meta API

**Pull request closed without merging**